packages/data-stores/src/domain-suggestions/reducer.ts

```diff
@@ -3,24 +3,71 @@
  */
 import type { Reducer } from 'redux';
 import { combineReducers } from '@wordpress/data';
+import type { TimestampMS } from 'wp-calypso-client/types';
 
 /**
  * Internal dependencies
  */
 import type { DomainSuggestion, DomainCategory } from './types';
 import type { Action } from './actions';
+import { DataState } from './constants';
 import { stringifyDomainQueryObject } from './utils';
 
-const domainSuggestions: Reducer< Record< string, DomainSuggestion[] | undefined >, Action > = (
-	state = {},
+export interface DomainSuggestionState {
+	state: DataState;
+	data: Record< string, DomainSuggestion[] | undefined >;
+	errorMessage: string | null;
+	lastUpdated: TimestampMS;
+	pendingSince: TimestampMS | undefined;
+}
```

Comment:
This is a simplification of what's in `http-data` which loses some important properties. For comparison:

https://github.com/Automattic/wp-calypso/blob/b9523f1420982eef9f659e2a04fb4b5efd065dfd/client/state/data-layer/http-data.ts#L17-L41

There's a practice often referred to as "making invalid states unrepresentable" which the http-data structure strives for but is lost here. [Here's an example](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nASigJgVwGMUo4BfOAMyghDgHIo98APegbgChRJYM5CtAEbAAdihwFiUAM7kqNOo2ZsunJKjgAVUChkwAhuACyAZTgBebChj4oorchQAeABQaUESnAAiBmCgAlADa9KIQAO70ALoAfJwoLLzwKKL4dH6GpoYBGJxwBXAAYgbAADZ2aFb0lKUVTPQANPmFAAqpuGIA5pYMqKKdol1NLQWmRMQyctUyE3oyI4VwAKqiYsAwwAZlwABeeL30+Gtrm9t7eCNknNyiAVC1xNYyEHbEmQbOWo1wAKKxeSW+n8KAAXL5-AZsiCuEtcJDwVo4AAfODHXAoShiPCwwokGhQcG-FFogaY7G4XEFMoGfTLMDwgK4RG6YEmUxUuD9QZdUxiYgskB6Qzskno8niSmca7qJzPV5QYhfXrHADW4Qioh+xKsao1ogBFlGJNcOBebxQHy+2oBADJ+MCAuCPtCAgA6VbrM47faUuCMgzg8VYyXsOD46BBskhnFcjrdPmiAWkjExv1kQLG1GmvQK96Q61-O0OnJgiFZUtu9oDbphiOElMS2PchP8ss6IVssBmciZpbZs15y0F75FuD2zCOssuyslcqVOtQAmIn4toaJ5PBim9rNwHPmxXDwyF-7jksg52Q10oN3jQiTGRhgMr8NLyONtNhte8ttR1PbjM1A8eULQAfSKCAIBgABxY5lV1UR1UiLU-hVRD9UNQFCinS8KxhY1n2sQQoFwZx9CgbofnI7oxWjClYk5etwWooYSTSMoyk5Gk6QZEFmW0VkRW7DljW-Dd20EoxhNo-9Q2lG4AHoFO0AALYA5FwCA9FEeh4BAAxVTQGRUmMt1VLQYCZEQfQUDoJCIjkGAVP8MyAHUUHoAA3NB9IxOADFEOAxE885cDgKdSSYMAmGMu4DCEMob04QRRH0FYTg2LYfTwVyNhUj5wUHC1enQY0cPLKFK09U4souXBmjhBF+BYcF6C0NS5BkFTXjKXBRAAHV0uAhDQbYykiS5yB+JS4AAYWgJhCHgYD62NJiGBUlAOIgOAImgXrFkKbiYHpRk8HBABGAAGG6boawoxN-OBrqu5oZRm9r1P9LSZB0vSDKMkyUB+IR8GWzaEDlKybLoL6ZH8dSsTwZKIFS+Bqsy85fVypyPnAyCYOOQrczAiCoNg0QyJgCihkw0qgVLXDKpBD0Mu9Or7oKQjMBahhPs67r8F6gahpG-ztoiSayE519lw2rbxt2-bcEO6laRO3imUu267tE+N1yel63qAA):

```ts
// Let's assume http-data Resource mentioned above.

// This doesn't make sense. The type system knows that. We've made an invalid state unrepresentable.
const UninitializedWithData: Resource = {
    state: DataState.Uninitialized,
    data: { x: "This shouldn't be allowed" },  // Correct type error 👍
    error: undefined,
    lastUpdated: -Infinity,
    pendingSince: undefined,
}

// Based on DomainSuggestionState here
// The single interface with union type members is permissive allowing invalid states to be built as we'll see below
interface Resource_FootGun {
    state: DataState;
    data: 'string';
    error: string | null;
    lastUpdated: TimestampMS;
    pendingSince: TimestampMS | undefined;
}

// This doesn't make sense, but the type system is satisfied
const UninitializedWithData_FootGun: Resource_FootGun = {
    state: DataState.Uninitialized,
    data: { x: 'This shouldn\'t be allowed' },  // We shouldn't have data in an uninitialized resource 😕
    error: undefined,
    lastUpdated: -Infinity,
    pendingSince: undefined,
}
```

That's not to say we need to use the `http-data` types verbatim, but I'd like to see:
- The higher level `Resource` types are shared across this level. They could be parameterized on the type of data and error [like in my example](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nASigJgVwGMUo4BfOAMyghDgHIo98APegbgChRJYM5CtAEbAAdihwFiUAM7kqNOo2ZsunJKjgAVUChkwAhuACyAZTgBebChj4oorchQAeABQaUESnAAiBmCgAlADa9KIQAO70ALoAfJwoLLzwKKL4dH6GpoYBGJxwBXAAYgbAADZ2aFb0lKUVTPQANPmFAAqpuGIA5pYMqKKdol1NLQWmRMQyctUyE3oyI4VwAKqiYsAwwAZlwABeeL30+Gtrm9t7eCNknNyiAVC1xNYyEHbEmQbOWo1wAKKxeSW+n8KAAXL5-AZsiCuEtcJDwVo4AAfODHXAoShiPCwwokGhQcG-FFogaY7G4XEFMoGfTLMDwgK4RG6YEmUxUuD9QZdUxiYgskB6Qzskno8niSmca7qJzPV5QYhfXrHADW4Qioh+xKsao1ogBFlGJNcOBebxQHy+2oBADJ+MCAuCPtCAgA6VbrM47faUuCMgzg8VYyXsOD46BBskhnFcjrdPmiAWkjExv1kQLG1GmvQK96Q61-O0OnJgiFZUtu9oDbphiOElMS2PchP8ss6IVssBmciZpbZs15y0F75FuD2zCOssuyslcqVOtQAmIn4toaJ5PBim9rNwHPmxXDwyF-7jksg52Q10oN3jQiTGRhgMr8NLyONtNhte8ttR1PbjM1A8eULQAfSKCAIBgABxY5lV1UR1UiLU-hVRD9UNQFCinS8KxhY1n2sQQoFwZx9CgbofnI7oxWjClYk5etwWooYSTSMoyk5Gk6QZEFmW0VkRW7DljW-Dd20EoxhNo-9Q2lG4AHoFO0AALYA5FwCA9FEeh4BAAxVTQGRUmMt1VLQYCZEQfQUDoJCIjkGAVP8MyAHUUHoAA3NB9IxOADFEOAxE885cDgKdSSYMAmGMu4DCEMob04QRRH0FYTg2LYfTwVyNhUj5wUHC1enQY0cPLKFK09U4souXBmjhBF+BYcF6C0NS5BkFTXjKXBRAAHV0uAhDQbYykiS5yB+JS4AAYWgJhCHgYD62NJiGBUlAOIgOAImgXrFkKbiYHpRk8HBABGAAGG6boawoxN-OBrqu5oZRm9r1P9LSZB0vSDKMkyUB+IR8GWzaEDlKybLoL6ZH8dSsTwZKIFS+Bqsy85fVypyPnAyCYOOQrczAiCoNg0QyJgCihkw0qgVLXDKpBD0Mu9Or7oKQjMBahhPs67r8F6gahpG-ztoiSayE519lw2rbxt2-bcEO6laRO3imUu267tE+N1yel63qAA).
- A few helper functions can help build the different states. The uninitialized state, for example, is always the same. The helper functions seem like they'd make sense for building action creators.
- Use the type system to make invalid states unrepresentable. We should have a union of better looking types.